### PR TITLE
SQL: Fix issue with field names containing "."

### PR DESCRIPTION
**Pull request by matriv**
Adjust FieldExtractor to be handle fields which contain `.` in their
name, regardless where they fall in in the document hierarchy. E.g.:

```
{
  "a.b": "Elastic Search"
}

{
  "a": {
    "b.c": "Elastic Search"
  }
}

{
  "a.b": {
    "c": {
      "d.e" : "Elastic Search"
    }
  }
}
```

Fixes: #37128

### CODE REVIEW

**Comment by matriv** [comment]
I tried to make use of the `XContentMapValues` methods `extractRawValues` and `filter` to replace the "ugly" code in the `FieldExtractor` but this seems not possible as the `extractRawValues` doesn't distinguish between a missing entry in the map and a `null` entry. For example if we have an Array `[null, null]` then the `extractRawValues` will return an empty List which prevents us from handling it correctly and throwing the "Arrays not supported" exception.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java [hunk 1]
```diff
@@ -144,19 +145,44 @@ Object extractFromSource(Map<String, Object> map) {
...
+                    }
+                }
             } else {
                 throw new SqlIllegalArgumentException("Cannot extract value [{}] from source", fieldName);
             }
         }
         return unwrapMultiValue(value);
     }
 
+    private Tuple<Object, Integer> extractAsDottedField(Map<String, Object> map, int idx, String node) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: I'm thinking to remove this extracted code back to the body of `extractFromSource()` as now we create an extra `Tuple` and `Integer` object which happens for every document returned. What do you think?

Also the code might be a bit ugly but I wanted to avoid calling recursively a method (with every combination of `a.b`, `a.b.c`, etc.) Happy to hear better suggestions.

**Comment by matriv** [comment]
run gradle build tests 2

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java [hunk 2]
```diff
@@ -39,7 +40,7 @@
      */
     private static String[] sourcePath(String name, boolean useDocValue, String hitName) {
         return useDocValue ? Strings.EMPTY_ARRAY : Strings
-                .tokenizeToStringArray(hitName == null ? name : name.substring(hitName.length() + 1), ".");
+            .tokenizeToStringArray(hitName == null ? name : name.substring(hitName.length() + 1), ".");
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: There's definitely some formatting differences here.
Anyway, make sure you set your IDE to format only the lines that were modified as that prevents unnecessary changes like the above.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: +1. It looks like a small method and while it might be inlined the extra `Tuple/Integer` are boiler-plate.
If the method gets unrolled, both the index and the found value will be available without wrapping/boxing.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java [hunk 3]
```diff
@@ -178,8 +204,8 @@ public boolean equals(Object obj) {
         }
         FieldHitExtractor other = (FieldHitExtractor) obj;
         return fieldName.equals(other.fieldName)
-                && hitName.equals(other.hitName)
-                && useDocValue == other.useDocValue;
+            && hitName.equals(other.hitName)
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Again with the formatting...

**Review by costin — APPROVED** [review]
LGTM.
I like the lack of virtual calls - the structure is small enough that we can extract the values through loops. +1 on unrolling the method so there's no additional method call.

Regarding tests, I would add a few more examples such as : `a.b { c { d.e } }` `a.b.c { d { e }` and `{a.b { c.d.e} }` just to be on the safe side.

**Comment by costin** [comment]
The random is fine :)

**Review by astefan — APPROVED** [review]
LGTM. Nice, elegant solution.

**Review by imotov — COMMENTED** [review]
I think we can improve it a bit.

```
DELETE test

PUT /test/blah/1
{
  "name.full": "Shane Connelly",
  "name.parts.first": "Shane"
}

PUT /test/blah/2
{
  "name": {
    "full": "Shane Connelly"
  },
  "name.parts": {
    "first": "Shane"
  }
}

POST /_xpack/sql?format=txt
{
  "query": "DESCRIBE test"
}

POST /_xpack/sql?format=txt
{
  "query": "select * from test"
}

GET /test/_search
{
  "_source": {
    "includes": "name.parts.first"
  }
}
```

**Comment by matriv** [comment]
@imotov I don't get your comment, how to improve it? in your example
```
POST /_xpack/sql?format=txt
{
  "query": "select * from test"
}
```
returns:
```
   name.full   |name.parts.first
---------------+----------------
Shane Connelly |Shane
Shane Connelly |Shane
```

**Comment by matriv** [comment]
I get:
```
   name.full   |name.parts.first
---------------+----------------
Shane Connelly |Shane
Shane Connelly |Shane
Shane Connelly |Shane II
``` :-)

**Comment by matriv** [comment]
Hm, I cleaned and recompiled and I get the null too, will check it, thx!

**Comment by matriv** [comment]
@imotov @costin @astefan Adjusted the code to handle multiple entries in the map with common prefixes. Please take another look.

Many thanks @imotov for catching that!

**Comment by matriv** [comment]
Now the query yields:
```
   name.full   |name.parts.first
---------------+----------------
Shane Connelly |Shane
Shane Connelly |Shane
Shane Connelly |Shane I
```

so it's `Shane I` and not `Shane II` since we're searching for the longer path in the map that yields non-null value. What do you think?

**Comment by costin** [comment]
How come the `null` was not caught by the tests before? What was the issue and what's the difference this time?
Regarding the test, I think there should be an exception when there are more than one value for the same field.
That is `{ a.b { c } }` is literally the same field as `{ a { b.c } }` if a document has definition for both fields it means it provides an array `{ a { b { c } } }  = [val1, val2]`.
So as it goes down on each level, the method should remember how many matching paths were found and determine whether there's a conflict.

**Comment by matriv** [comment]
@costin Check this: https://github.com/elastic/elasticsearch/pull/37364/files#diff-9aaee9be08445653bba7407b9f2b5ca3R312
The previous tests only had a single entry in the map.

Will make the changes to catch the same path(s)/different hierarchy(ies) and throw an exception about not supporting arrays (multi-values).

**Comment by matriv** [comment]
The difference now is here: https://github.com/elastic/elasticsearch/pull/37364/files#diff-6d292f4ab3ddfc415649c328fd2faefeR157
So we don't stop once a non-null value is found in the map but we keep walking the path and holding the latest non-null value found.

**Comment by matriv** [comment]
Pushed the new approach to handle multiple values for the same path but different hierarchies.

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by matriv** [comment]
@elasticmachine  run default-distro

**Comment by matriv** [comment]
run default distro tests

**Comment by matriv** [comment]
run default distro tests

**Comment by matriv** [comment]
run gradle build tests 1

**Review by imotov — APPROVED** [review]
LGTM. Thanks!

**Comment by matriv** [comment]
@imotov thanks for checking and catching this.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java [hunk 4]
```diff
@@ -141,17 +143,47 @@ private Object unwrapMultiValue(Object values) {
...
-            } else if (first || value instanceof Map) {
-                first = false;
-                value = ((Map<String, Object>) value).get(node);
-            } else {
-                throw new SqlIllegalArgumentException("Cannot extract value [{}] from source", fieldName);
+        Object value = null;
+
+        // Used to avoid recursive method calls
+        // Holds the sub-maps in the document hierarchy that are pending to be inspected.
+        LinkedList<Map<String, Object>> queue = new LinkedList<>();
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: minor nitpick - use the `Deque` instead of `LinkedList`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java [hunk 5]
```diff
@@ -141,17 +143,47 @@ private Object unwrapMultiValue(Object values) {
...
+            // e.g.: If the path == "a.b.c.d" and the idx == 0, we need to check the current subMap against the keys:
+            //       "b", "b.c" and "b.c.d"
+            StringJoiner sj = new StringJoiner(".");
+            for (int i = idx + 1; i < path.length; i++) {
+                sj.add(path[i]);
+                Object node = subMap.get(sj.toString());
+                if (node instanceof Map) {
+                    // Add the sub-map to the queue along with the current path index
+                    queue.add((Map<String, Object>) node);
+                    idxQueue.add(i);
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Why not add i+1 to the queue since the next loop should start from the next position?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Since the index and the Map are associated, how about using only one `Deque` which holds a `Tuple` instead of two `Deque`:

```
Deque<Tuple<Map<String, Object> index>>` queue = ...

if (node instanceof Map) {
   queue.add(new Tuple<>(node, Integer.valueOf(i));
}

```

**Review by costin — APPROVED** [review]
Left some minor comments. Let's get this in.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Actually, I had those 2 before while implementing. :-)

I chose `LinkedList` as we don't need any special operations of queue (we don't care about the order of dequeuing). I've just read about the better performance of `ArrayDequeue` over `LinkedList` and I'll change.

For the `Tuple` I had it as you said but I changed to a separate structure that holds integers to avoid the instantiation of `Tuple` objects on top of the Integer boxing. What do you think?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 5 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: no reason, I just preferred to start the `i` index of the for loop from prevPosition + 1 just because personally I see it more "visible". Would you prefer it your way?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: And after benchmarking: https://gist.github.com/matriv/07e5909a49bed9794f350f458a0b2c60
The results show:
```
Benchmark                      Mode  Cnt       Score       Error  Units
MyBenchmark.testArrayDequeue  thrpt   25  608384,000 ± 7893,014  ops/s
MyBenchmark.testLinkedList    thrpt   25  581316,884 ±  8028,331  ops/s
```

Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Autoboxing already happens and I wouldn't worry to much about it considering the depth is not that big. Same for `Linked` vs `Array` (in general arrays are faster except for inserting in the middle as that requires resizing/copying at which the linked structure excels).
I think the `Tuple` makes the code a bit more compact and safe (the queues cannot get out of sync) and more readable/simple code always trumps optimization (especially micro ones as here).

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 5 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: No.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java]: Just FYI regarding the Tuples:
https://gist.github.com/matriv/3d2557ae1621e99a3c9505b5e4e6998f
and the result:
```
Benchmark                      Mode  Cnt       Score      Error  Units
MyBenchmark.testNoTuples      thrpt   25  601543,170 ± 7859,704  ops/s
MyBenchmark.testTuples        thrpt   25  573953,828 ± 5987,749  ops/s
```

**Comment by matriv** [comment]
Backported to `6.6` with 874b5ecba3e95d654dd0a04b7d19f9e9f050ba87

**Comment by matriv** [comment]
Backported to `6.x` with c72bd51c80f8835d7f9d1e218358862211589d54

**Comment by matriv** [comment]
Backported to `6.x` with fd3701b8c32b251b514c9d6cd98f2837dded7fa4

**Comment by imotov** [comment]
@matriv I am sorry. It looks like I missed another edge case during review and immediately got bitten by it in geosql, which actually can extract objects from source in some cases. I opened #37502 to fix it.
